File: websocket/code/pong/CanvasObject.py
import math
import json

class CanvasObject:
    def __init__(self, obj):
        # TODO: crear un diccionario que nazca de parsear la string JSON obj
        self.color = "white"
        self.id = "generic"
        self.type = "generic"
        self.x = 0
        self.y = 0
        self.dirX = 0
        self.dirY = 0
        self.speed = 4
        self.is_moving = False
        self.width = 0
        self.height = 0
        self.canvas_width = 800
        self.canvas_height = 400
        self.pk = -42

        if 'color' in obj and type(obj['color]'] is str):
            self.color = str(obj['color'])
        else:
            self.color = "white"

        if 'id' in obj and type(obj['id']) is str:
            self.id = str(obj['id'])
        else:
            self.id = "generic"

        if 'type' in obj and type(obj['type']) is str:
            self.type = str(obj['type'])
        else:
            self.type = "generic"

        if 'x' in obj and (type(obj['x']) is float or type(obj['x']) is int):
            self.x = float(obj['x'])
        else:
            self.x = 0

        if 'y' in obj and (type(obj['y']) is float or type(obj['y']) is int):
            self.y = float(obj['y'])
        else:
            self.y = 0

        if 'dirX' in obj and (type(obj['dirX']) is float or type(obj['dirX']) is int):
            self.dirX = float(obj['dirX'])
        else:
            self.dirX = 0

        if 'dirY' in obj and (type(obj['dirY']) is float or type(obj['dirY']) is int):
            self.dirY = float(obj['dirY'])
        else:
            self.dirY = 0

        if 'speed' in obj and (type(obj['speed']) is float or type(obj['speed']) is int):
            self.speed = float(obj['speed'])
        else:
            self.speed = 4

        if 'is_moving' in obj and type(obj['is_moving'] is bool):
            self.is_moving = obj['is_moving']
        else:
            self.is_moving = False

        if 'width' in obj and (type(obj['width']) is float or type(obj['width']) is int):
            self.width = float(obj['width'])
        else:
            self.width = 0

        if 'height' in obj and (type(obj['height']) is float or type(obj['height']) is int):
            self.height = float(obj['height'])
        else:
            self.height = 0

        if 'canvas_width' in obj and (type(obj['canvas_width']) is float or type(obj['canvas_width']) is int):
            self.canvas_width = float(obj['canvas_width'])
        else:
            self.canvas_width = 800

        if 'canvas_height' in obj and (type(obj['canvas_height']) is float or type(obj['canvas_height']) is int):
            self.canvas_height = float(obj['canvas_height'])
        else:
            self.canvas_height = 400

        if 'font' in obj and type(obj['font'] is str):
            self.font = str(obj['font'])
        else:
            self.font = "42px Arial"


        # xy1---------xy2
        # |            |
        # |     xy     |
        # |            |
        # xy3---------xy4

        self.point_x1 = self.x - self.width / 2
        self.point_y1 = self.y - self.height / 2

        self.point_x2 = self.x + self.width / 2
        self.point_y2 = self.y - self.height / 2

        self.point_x3 = self.x - self.width / 2
        self.point_y3 = self.y + self.height / 2

        self.point_x4 = self.x + self.width / 2
        self.point_y4 = self.y + self.height / 2
    
    def serialize(self):
        return json.dumps({
            'id': self.id,
            'type': self.type,

            'x': self.x,
            'y': self.y,
            
            'width': self.width,
            'height': self.height,

            'color': self.color,
        }) 

    def objectHits(self, object):
        horizontal_check = True
        vertical_check = True

        horizontal_proyection_center = math.fabs(self.x - object.x)
        horizontal_proyection_obj1 = math.fabs(self.x - self.point_x2)
        horizontal_proyection_obj2 = math.fabs(object.x - object.point_x2)

        if (horizontal_proyection_obj1 + horizontal_proyection_obj2) < horizontal_proyection_center:
            horizontal_check = False

        vertical_proyection_center = math.fabs(self.y - object.y)
        vertical_proyection_obj1 = math.fabs(self.y - self.point_y2)
        vertical_proyection_obj2 = math.fabs(object.y - object.point_y2)

        if (vertical_proyection_obj1 + vertical_proyection_obj2) < vertical_proyection_center:
            vertical_check = False

        return horizontal_check and vertical_check

    def resolveHit(self, canvas_object):
        pass

    def slide(self, dirX, dirY):
        length = math.hypot(dirX, dirY)
        if (length > 0):
            dirX /= length
            dirY /= length

        self.x += self.speed * dirX
        self.y += self.speed * dirY
        
        self.dirX = dirX
        self.dirY = dirY
        
        self.recalculateHitbox()
        self.keepInsideCanvas()

    def moveTo(self, x, y):
        self.x = x
        self.y = y

        self.recalculateHitbox()
        self.keepInsideCanvas()

    def update(self, canvas_objects):
        self.slide(self.dirX, self.dirY)
        for obj in canvas_objects:
            if obj != self and self.objectHits(obj):
                self.resolveHit(obj);
    
    def recalculateHitbox(self):
        self.point_x1 = self.x - self.width / 2;
        self.point_y1 = self.y - self.height / 2;

        self.point_x2 = self.x + self.width / 2;
        self.point_y2 = self.y - self.height / 2;

        self.point_x3 = self.x - self.width / 2;
        self.point_y3 = self.y + self.height / 2;

        self.point_x4 = self.x + self.width / 2;
        self.point_y4 = self.y + self.height / 2;

    def keepInsideCanvas(self):
        # TODO: el canvas como te lo van a pasar, solo necesitas el tamano, pero tendra que pasarse en el json del constructor
        if self.point_x1 < 0:
            self.moveTo((self.point_x2 - self.point_x1) / 2, self.y)
        elif self.point_x2 > self.canvas_width:
            self.moveTo(self.canvas_width - (self.point_x2 - self.point_x1) / 2, self.y)

        if self.point_y1 < self.canvas_height / 8:
            self.moveTo(self.x, (self.canvas_height / 8) + ((self.point_y3 - self.point_y1) / 2));
        elif self.point_y3 > self.canvas_height:
            self.moveTo(self.x, self.canvas_height - ((self.point_y3 - self.point_y1) / 2));

File: nginx-files/static-files/pong/CanvasObject.js
"use strict";
import * as THREE from 'three';

class CanvasObject {
/**
 * @param obj individual object of parsed with JSON.parse()
 * @param canvas instance of the canvas
 * @param context instance of the context
 */
constructor(obj, canvas, scene) {
	this.canvas = canvas;

	this.color = "white";
	this.id = "generic";
	this.type = "generic";
	this.x = 0;
	this.y = 0;
	this.dirX = 0;
	this.dirY = 0;
	this.speed = 4;
	this.is_moving = false;
	this.width = 0;
	this.height = 0;

	for (let i in obj)
		this[i] = obj[i];

	/*
	** xy1---------xy2
	** |            |
	** |     xy     |
	** |            |
	** xy3---------xy4
	*/
	this.point_x1 = this.x - this.width / 2;
	this.point_y1 = this.y - this.height / 2;

	this.point_x2 = this.x + this.width / 2;
	this.point_y2 = this.y - this.height / 2;

	this.point_x3 = this.x - this.width / 2;
	this.point_y3 = this.y + this.height / 2;

	this.point_x4 = this.x + this.width / 2;
	this.point_y4 = this.y + this.height / 2;

	this.geometry = new THREE.BoxGeometry(this.width, this.height, 20);
	this.material = new THREE.MeshPhongMaterial()
	this.mesh = new THREE.Mesh(this.geometry, this.material);
	this.mesh.position.x = this.x;
	this.mesh.position.y = - this.y;
	scene.add(this.mesh);
}

animate(obj) {
	this.mesh.position.x = obj.x;
	this.mesh.position.y = - obj.y;
}

/**
 * @brief puts the hitbox in the canvas
 */
renderHitBox() {
	this.context.strokeStyle = "magenta";
	this.context.beginPath();
	this.context.lineWidth = "1";
	this.context.rect(this.point_x1, this.point_y1, this.width, this.height);
	this.context.closePath();
	this.context.stroke();
}

/**
 * @brief checks if a given canvasObject is in contact
 * @param {CanvasObject} object to check
 * @returns true if is in contact, false otherwise
 */
objectHits(object) {
	let horizontal_check = true;
	let vertical_check = true;

	const horizontal_proyection_center = Math.abs(this.x - object.x);
	const horizontal_proyection_obj1 = Math.abs(this.x - this.point_x2);
	const horizontal_proyection_obj2 = Math.abs(object.x - object.point_x2);
	
	if ((horizontal_proyection_obj1 + horizontal_proyection_obj2) < horizontal_proyection_center)
		horizontal_check = false;

	const vertical_proyection_center = Math.abs(this.y - object.y);
	const vertical_proyection_obj1 = Math.abs(this.y - this.point_y2);
	const vertical_proyection_obj2 = Math.abs(object.y - object.point_y2);

	if ((vertical_proyection_obj1 + vertical_proyection_obj2) < vertical_proyection_center)
		vertical_check = false;

	return horizontal_check && vertical_check;
}

/**
 * @brief moves the object in (x, y), it DOESN'T move to global (x,y),
 * it moves from current position in (x, y)
 * @param {int} dirX direction in x axis
 * @param {int} dirY direction in y axis
*/
slide(dirX, dirY) {
	let length = Math.hypot(dirX, dirY);
	if (length > 0) {
	    dirX /= length;
	    dirY /= length;
	}

	this.x += this.speed * dirX;
	this.y += this.speed * dirY;
	
	this.dirX = dirX;
	this.dirY = dirY;
	
	this.recalculateHitbox();
	this.keepInsideCanvas();
}

/**
 * @brief moves the object to global(x, y)
 * @param {int} x position in x
 * @param {int} y position in y 
 */
moveTo(x, y) {
	this.x = x;
	this.y = y;	
	this.recalculateHitbox();
	this.keepInsideCanvas();
}

/**
 * @brief updates the info of the object, it includes movement and collisions
 * @param {CanvasObject[]} canvas_objects objects in the current canvas
 */
update(canvas_objects) {
	this.slide(this.dirX, this.dirY);
	for (let i = 0; i < canvas_objects.length; i++) {
		if (canvas_objects[i] != this && this.objectHits(canvas_objects[i]))
			this.resolveHit(canvas_objects[i]);
	}
}

/**
 * @brief uptades the object state when is hitted
 * @param {CanvasObject} canvas_object object in contact
 */
resolveHit(canvas_object) {
}

/**
 * @brief recalculates the hitbox points with the current (x, y)
 */
recalculateHitbox() {
	this.point_x1 = this.x - this.width / 2;
	this.point_y1 = this.y - this.height / 2;

	this.point_x2 = this.x + this.width / 2;
	this.point_y2 = this.y - this.height / 2;

	this.point_x3 = this.x - this.width / 2;
	this.point_y3 = this.y + this.height / 2;

	this.point_x4 = this.x + this.width / 2;
	this.point_y4 = this.y + this.height / 2;
}

/**
 * @brief calculate if the object is outside the canvas, and moves it to
 * the border
 */
keepInsideCanvas() {
	if (this.point_x1 < 0)
		this.moveTo((this.point_x2 - this.point_x1) / 2, this.y)
	else if (this.point_x2 > this.canvas.width)
		this.moveTo(this.canvas.width - (this.point_x2 - this.point_x1) / 2, this.y)

	if (this.point_y1 < this.canvas.height / 8)
		this.moveTo(this.x, (this.canvas.height / 8) + ((this.point_y3 - this.point_y1) / 2));
	else if (this.point_y3 > this.canvas.height)
		this.moveTo(this.x, this.canvas.height - ((this.point_y3 - this.point_y1) / 2));
}

/**
 * @param {float} x x value
 * @param {float} y y value
 * @returns the direction (in form of an angle in degrees) of the given vector
 */
vector_dir(x, y)  {
	const a = Math.atan(Math.abs(y) / Math.abs(x)) * (180 / Math.PI);

	if (x >= 0 && y >= 0)
		return a;
	if (x <= 0 && y > 0)
		return 180 - a;
	if (x <= 0 && y <= 0)
		return 180 + a;
	if (x > 0 && y < 0)
		return 360 - a;
}

/**
 * @param {int} angle angle value in degrees
 * @returns a string with the corresponding North, East, South or West from a given angle
 */
position_view(angle) {
	if (angle >= 45 && angle < 135)
		return "North"
	if (angle >= 135 && angle < 225)
		return "East"
	if (angle >= 225 && angle < 315)
		return "South";
	return "West";
}
}

export { CanvasObject };
